rfcs/002-references-and-pointers.md

```diff
@@ -0,0 +1,332 @@
+```
+
+**To understand this RFC, stop thinking of `mut` parameters as pointers, this is merely an implementation detail and should not be exposed to V.**
+
+**`mut` parameters are just variables whos origin lies outside of the function, as far as users are concerned, it is not a pointer and should not act like one.**
+
+Mutable non-pointer parameters should function as references in C++, or an `inout` parameter in other languages. 
+
+```vlang
+struct Struct {
+	/* ... */
+}
+
+fn take_struct(mut v Struct) {
+	ptr := &Struct(/* ... */)
+
+	// no pointer arithmetic, never nil.
+	// it's just a variable in the sense of the function, not a pointer here.
+	v = Struct{/* ... */}
+
+	field := v.field // auto "deref" on fields, this is normal
+	v = *ptr         // need to deref here, overwrite memory
+}
+```
+
+Below is a pointer, it functions equivalently with mutability however assignments are different. This is the current behavior in V.
+
+```vlang
+struct Struct {
+	/* ... */
+}
+
+fn take_struct(mut v &Struct) {
+	ptr := &Struct(/* ... */)
+
+	// pointer arithmetic, may be nil in unsafe contexts.
+	// it's a pointer, overwrite memory using *
+	*v = Struct{/* ... */}
+
+	field := v.field // auto deref on fields, this is normal
+	v = ptr          // assigning integers to integers..
+}
+```
+
+With the above, it's understood that the second example now explicitly takes a pointer, and the first example is just a variable that can be edited.
+
+---
+
+Instead of thinking of `mut params` as pointers, we can be explicit here.
+
+The parameter of the `fn send` is mutable, it's a mutable binding to a variable outside of the function. We can take the address of it with the `&` operator, and it's the address of that outside variable.
+
+```vlang
+struct AA {/* ... */}
+
+// the `mut AA` doesn't make sense for external C functions.
+// fn C.external(mut AA)
+
+fn C.external(mut &AA)
+
+fn send(mut value AA) {
+	C.external(&value)
+}
+```
+
+You can't cast a `mut param` to a pointer, because it isn't a pointer. You must take the address of it.
+
+No need to explicitly construct a pointer value, if you allow V to promote it's allocation you won't leave auto-deref territory.
+
+You shouldn't even think in "promotion of allocations." At the end of the day a variable is a variable, and pointers are hidden.
+
+```vlang
+struct AA {/* ... */}
+
+fn return_aa(value AA) &AA {
+	mut a := AA{}
+	a.field = 10
+	a = value     // "auto deref" -- this isn't a pointer to users
+	return a
+}
+```
+
+**What about immutable `&T` ?**
+
+When a `T` is immutable, it doesn't matter if it is passed by value or by reference. There is zero difference in underlying assembly code. Immutability plays a big part here.
+
+**90% of the time, you will never use a `&T` in V code. It should be discouraged.**
+
+If you're working with linked lists, need to set pointer values, that's fine. However, if you're working with an immutable receiver it doesn't matter.
```

Comment:
Doesn't it matter if the struct is big? It's less effort to pass a pointer than copying the whole struct. Isn't it?

Reply:
It makes zero difference, and having V users think about it is pointless.

Reply:
So you would pass it as a reference always? I don't get it.

Reply:
> So you would pass it as a reference always? I don't get it.

In generated code, all structures >= 16 bytes will always be passed as a pointer. So, passing by value and by pointer functions exactly the same if the values are immutable.

Reply:
Ah ok, but I thought passing >=16 bytes => pointer mechanism shouldn't be done by the compiler anymore?
Maybe this could be specified in this rfc?

Reply:
This is done when generating code, machine code. It is not up to V. This just an implementation detail of how parameter passing works in most computer architectures, and should be out of the scope of any V user. I am simply demonstrating that a pass by value or pass by reference, when immutable, are functionally equivalent.